pkg/bridge/ai/provider/openai/provider.go

```diff
@@ -0,0 +1,270 @@
+package openai
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+
+	_ "github.com/joho/godotenv/autoload"
+	"github.com/yomorun/yomo/ai"
+	"github.com/yomorun/yomo/core/ylog"
+)
+
+const APIEndpoint = "https://api.openai.com/v1/chat/completions"
+
+var fns sync.Map
+
+// Message
+type ChatCompletionMessage struct {
+	Role    string `json:"role"`
+	Content string `json:"content"`
+	// - https://github.com/openai/openai-python/blob/main/chatml.md
+	// - https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+	Name string `json:"name,omitempty"`
+	// MultiContent []ChatMessagePart
+	// For Role=assistant prompts this may be set to the tool calls generated by the model, such as function calls.
+	ToolCalls  []ai.ToolCall `json:"tool_calls,omitempty"`
+	ToolCallID string        `json:"tool_call_id,omitempty"`
+}
+
+// RequestBody is the request body
+type ReqBody struct {
+	Model    string                  `json:"model"`
+	Messages []ChatCompletionMessage `json:"messages"`
+	Tools    []ai.ToolCall           `json:"tools"` // chatCompletionTool
+	// ToolChoice string    `json:"tool_choice"` // chatCompletionFunction
+}
+
+// Resp is the response body
+type RespBody struct {
+	ID                string       `json:"id"`
+	Object            string       `json:"object"`
+	Created           int          `json:"created"`
+	Model             string       `json:"model"`
+	Choices           []RespChoice `json:"choices"`
+	Usage             RespUsage    `json:"usage"`
+	SystemFingerprint string       `json:"system_fingerprint"`
+}
+
+// RespMessage is the message in Response
+type RespMessage struct {
+	Role      string        `json:"role"`
+	Content   string        `json:"content"`
+	ToolCalls []ai.ToolCall `json:"tool_calls"`
+}
+
+// RespChoice is used to indicate the choice in Response by `FinishReason`
+type RespChoice struct {
+	FinishReason string                `json:"finish_reason"`
+	Index        int                   `json:"index"`
+	Message      ChatCompletionMessage `json:"message"`
+}
+
+// RespUsage is the token usage in Response
+type RespUsage struct {
+	PromptTokens     int `json:"prompt_tokens"`
+	CompletionTokens int `json:"completion_tokens"`
+	TotalTokens      int `json:"total_tokens"`
+}
+
+// OpenAIProvider is the provider for OpenAI
+type OpenAIProvider struct {
+	// APIKey is the API key for OpenAI
+	APIKey string
+	// Model is the model for OpenAI
+	// eg. "gpt-3.5-turbo-1106", "gpt-4-turbo-preview", "gpt-4-vision-preview", "gpt-4"
+	Model string
+}
+
+type connectedFn struct {
+	connID uint64
+	tag    uint32
+	tc     ai.ToolCall
+}
+
+func init() {
+	fns = sync.Map{}
```

Comment:
The initialization of `fns` is not necessary because the `fns` variable is available after declaring it with `var fns sync.Map`.

Reply:
I am considering put `fns` to `OpenAIProvider` as an instance property.